File: JavaScript/analyzer.js
/*
 * Это простой редактор JavaScript.
 *
 * Введите JavaScript, затем щёлкните правой кнопкой или выберите из меню Выполнить:
 * 1. Запустить, чтобы исполнить выделенный текст (Ctrl+R),
 * 2. Исследовать, чтобы вызвать для результата Инспектор Объектов (Ctrl+I), или,
 * 3. Отобразить, чтобы вставить результат в комментарий после выделения. (Ctrl+L)
 */"use strict";
var digits = '0123456789';
var letters = 'abcdefghijklmnopqrstuvwxyz';
letters += letters.toUpperCase();
letters += '_$';
var quotes = '\'"';
var alphabeth = letters + digits;
var spaces = ' \t';
var is_log = false;

function returnHash(str) {
    var i = 0;
    var h = {
    };
    for (i = 0;i < str.length;i++)
        h[str[i]] = true;
    return h;
}

digits = returnHash(digits);
letters = returnHash(letters);
quotes = returnHash(quotes);
alphabeth = returnHash(alphabeth);
spaces = returnHash(spaces);
dividers = {
    ";" : true, "," : true
};

function getDivider(string, beg) {
    var i = beg;
    var res = {
    };
    while (alphabeth[string[i]]) {
        i++;
    }
    res = string.slice(beg, i);
    return {value : res, index : i, type : 'divider', error : ''}

}

function getWord(string, beg) {
    var i = beg;
    var res = {
    }
    while (alphabeth[string[i]]) {
        i++;
    }
    res = string.slice(beg, i);
    return {value : res, index : i, type : 'word', error : ''}
}

function getInteger(string, beg) {
    var i = beg;
    var res = '';
    while (digits[string[i]]) {
        i++;
    }
    res = string.slice(beg, i);
    return {value : res, index : i, type : 'number', subtype : 'integer', error : ''}
}

function getLiteral(string, beg) {
    var i = beg;
    var quot = string[beg];
    var curr = string[beg];
    var prev = '';
    var quote_closed = false;
    var string_ended = false;
    var error = '';
    while (!((string_ended || quote_closed))) {
        if (is_log)
            console.log('getLiteral: ' + i + " |" + string[i] + "| " + string_ended + " " + quote_closed);
        prev = string[i];
        i++;
        var backslash = "\\";
        curr = string[i];
        if (curr == backslash) {
            i++;
            prev = curr;
        }
        else if ((curr == quot) && (prev != backslash)) {
            if (is_log)
                console.log("getLiteral:quotemode curr = " + curr + " prev = " + prev)
            quote_closed = true;
            i++
        }
        if (i >= string.length) {
            error = "literal not ended"
            string_ended = true;
        }

    }
    // of while
    var res = string.slice(beg, i);
    res.error = '';
    if (!quote_closed)
        res.error = 1;
    res = {
        value : res, index : i, type : 'literal'
    };
    //if (!quote_closed) 
    res.error = error;

    return res;
}

function getElements(string) {
    var i = 0, prev_index = 0;
    var elements = [];
    var result = {
        elements : elements, error : false
    };
    var res = '';

    while (i < string.length) {
        while (spaces[string[i]])
            i++;
        if (is_log)
            console.log("getElements:index of " + i + " letter is " + string[i]);
        prev_index = i;
        res = {
        };
        if (!spaces[string[i]]) {
            if (letters[string[i]]) {
                res = getWord(string, i);

            }
            else if (quotes[string[i]]) {
                res = getLiteral(string, i);

            }
            else if (digits[string[i]]) {
                res = getInteger(string, i);
            }
            else if (dividers[string[i]]) {
                res = getDividers(string, i);
            }

            if (res.error != '') {
                result.error = res.error;

            }

            if ((!res.hasOwnProperty('value'))) {
                result.error = true;
                result.error = 'undefined position at ' + i;
                return result;

            }
            elements.push(res);
            i = res.index;
        };

    }
    return result;
}

var str = 'mov eax, 01 "Landmard of \\\" Sir \\\"  " 233\"';
console.log(getElements(str))

/*
Exception: SyntaxError: expected expression, got ';'
@Scratchpad/1:37
*/

File: Ruby/analyzer.rb
#!/bin/env ruby
# encoding: utf-8

prm = ARGV;

def str_to_presenceHash(str, divider='')
  the_divider = '';
  the_divider = divider if divider;
  the_sym = str.split(the_divider).inject({}) {|r,x| r[x] = 1; r};
  the_sym
end

class Analyzer
  def initialize(params={})
    
    #default valeus
    @letters = (('a'..'z').to_a.join('') + ('A'..'Z').to_a.join('')) + '_$';
    @digits  = (0..9).to_a.join('');
    @quotes  = "'\"`~";
    @spaces  = "\t ";
    @reserved = 'mov goto add codesegmend begin end datasegment';
    @code  = '';
    
    @letters = params['letters'] if params['letters']
    @digits  = params['digits'] if params['digits']
    @quotes  = params['quotes'] if params['quotes']
    @spaces = params['spaces'] if params['spaces']
    @reserved = params['reserved'] if params['reserved']
    @code = File.open(params['filename'], 'r'){|read| read.read} if params['filename']
    
    @letters = str_to_presenceHash(@letters);
    @digits = str_to_presenceHash(@digits);
    @quotes = str_to_presenceHash(@quotes);
    @spaces = str_to_presenceHash(@spaces);
    @reserved = str_to_presenceHash(@reserved, ' ');
    
    @alphabeth = @letters.merge(@digits);
    @is_log = false;
  end
  
  def getWord(string, beg)

    i = beg;
    while @alphabeth[string[i]]
      i+=1;
      # puts "#{i}"
    end
    return {'value'=>string[beg..i-1], 'index'=>i, 'type'=>'word', 'error'=>''}
  end
  
  def getInteger(string, beg)
    i = 0
    while @digits[string[i]]
      i+=1;
    end
    return {'value'=>string[beg..i-1], 'index'=>i, 'type'=>'integer', 'error'=>''}
  end
  
  def getLiteral(string, beg)
    p "getLiteral:start" if @is_log
    quot = string[beg];
    backslash = '\\';
    curr = string[beg]
    prev = string[beg]
    string_ended = false;
    literal_closed = false;
    i = beg;
    while !string_ended & !literal_closed
      i+=1;
      curr = string[i];
      if curr == backslash
          prev = curr
          i+=1;
      end
      if curr == quot && prev != backslash
        i+=1   
        literal_closed = 1
      elsif i >= string.size
        string_ended = true
        
      end
      
    end #of while !string_ended & !literal_closed
    error = '';
    error = 'literal not closed' if string_ended && !literal_closed
    p "getLiteral:end" if @is_log
    return {'value'=>string[beg..i-1], 'index'=>i, 'type'=>'literal', 'error'=>''}
  end
  
  def getElements(string)

    result = {};
    result['elements'] = [];
    res = {};
    i = 0;
    len = string.size
    while i < len
      i+=1 if @spaces[string[i]]
      puts "index is #{i}" if @is_log

      if @letters[string[i]]
        res = getWord(string, i)
      elsif @digits[string[i]]
        res = getInteger(string, i)
      elsif @quotes[string[i]]
        res = getLiteral(string, i)
      end
      i = res['index'];
      result['elements'].push(res);
      if res['error'] !=''
  result['error'] = res.error;
  return result;
  
      end
      
    end
    return result;
  end
  
end

a = Analyzer.new()

text = a.getElements('Sir Walter Scott of "Abbotsworth and \"Lynd\" Castle"');
p text

